File: Sources/ForgeCore/RepositoryExperience.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

struct RepositoryOverview: Decodable, Sendable {
    struct License: Decodable, Sendable { let name: String; let spdxId: String? }
    let fullName: String
    let description: String?
    let defaultBranch: String
    let stargazersCount: Int
    let forksCount: Int
    let subscribersCount: Int?
    let openIssuesCount: Int
    let permissions: RepositorySettings.Permissions?
    let license: License?
}

struct PinnedRepository: Decodable, Identifiable, Sendable {
    let nameWithOwner: String
    let description: String?
    let stargazerCount: Int
    var id: String { nameWithOwner }
}

struct ProfileHighlights: Decodable, Sendable {
    struct Pins: Decodable, Sendable { let nodes: [PinnedRepository?] }
    let pinnedItems: Pins
    let isEmployee: Bool
    let isDeveloperProgramMember: Bool
    let isGitHubStar: Bool
    let isCampusExpert: Bool
    var badges: [String] {
        [(isEmployee, "GitHub Staff"), (isDeveloperProgramMember, "Developer Program"),
         (isGitHubStar, "GitHub Star"), (isCampusExpert, "Campus Expert")].filter(\.0).map(\.1)
    }
}

struct IssueFields {
    var title: String
    var body: String
    var assignees: [String] = []
    var labels: [String] = []
    var milestone: Int? = nil
    func payload() throws -> [String: Any] {
        let title = title.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !title.isEmpty, title.count <= 256, assignees.allSatisfy(GitHubAccount.validLogin), assignees.count <= 10,
              milestone == nil || milestone! > 0 else { throw GitHubError("Enter a title and valid issue metadata.") }
        var value: [String: Any] = ["title": title, "body": body]
        if !assignees.isEmpty { value["assignees"] = assignees }
        if !labels.isEmpty { value["labels"] = labels }
        if let milestone { value["milestone"] = milestone }
        return value
    }
}

struct IssueOption: Identifiable, Hashable, Sendable {
    let id: String
    let title: String
    var detail = ""
}

struct IssueOptionPage: Sendable { let items: [IssueOption]; let more: Bool; var cursor: String? = nil }

struct ReadmeDocument: Sendable {
    struct Heading: Identifiable, Sendable { let id: String; let title: String; let level: Int }
    let repository: Repository
    let sha: String
    let path: String
    let html: String
    let headings: [Heading]
    var baseURL: URL { var c = URLComponents(); c.scheme = "https"; c.host = "github.com"; c.path = "/\(repository.fullName)/blob/\(sha)/\(path)"; return c.url! }

    init(html: String, repository: Repository, sha: String, path: String) throws {
        guard sha.range(of: "^[a-fA-F0-9]{40}$", options: .regularExpression) != nil,
              RepositoryFile(name: "", path: path, sha: sha, type: "file", size: 0).safePath else { throw GitHubError("Invalid README revision or path.") }
        self.repository = repository; self.sha = sha; self.path = path
        var source = html.replacingOccurrences(of: #"\s+srcset\s*=\s*(?:"[^"]*"|'[^']*')"#, with: "", options: [.regularExpression, .caseInsensitive])
        let regex = try NSRegularExpression(pattern: #"\bsrc\s*=\s*(["'])(.*?)\1"#, options: [.caseInsensitive, .dotMatchesLineSeparators])
        var base = URLComponents(); base.scheme = "https"; base.host = "raw.githubusercontent.com"; base.path = "/\(repository.fullName)/\(sha)/\(path)"
        for match in regex.matches(in: source, range: NSRange(source.startIndex..., in: source)).reversed() {
            let raw = (source as NSString).substring(with: match.range(at: 2)).replacingOccurrences(of: "&amp;", with: "&")
            var replacement = ""
            if let url = URL(string: raw, relativeTo: base.url)?.absoluteURL.standardized, url.scheme == "https", url.user == nil, url.password == nil, url.port == nil || url.port == 443 {
                let rawPrefix = "/\(repository.fullName)/\(sha)/", webPrefix = "/\(repository.fullName)/raw/\(sha)/"
                let imagePath: String?
                if url.host == "raw.githubusercontent.com", url.path.hasPrefix(rawPrefix) { imagePath = String(url.path.dropFirst(rawPrefix.count)) }
                else if url.host == "github.com", url.path.hasPrefix(webPrefix) { imagePath = String(url.path.dropFirst(webPrefix.count)) }
                else { imagePath = nil }
                if let imagePath {
                    var local = URLComponents(); local.scheme = "forge-readme"; local.host = "image"; local.path = "/" + imagePath
                    replacement = local.url!.absoluteString
                } else { replacement = url.absoluteString }
            }
            source = (source as NSString).replacingCharacters(in: match.range, with: "src=\"\(Self.escape(replacement))\"")
        }
        let pattern = try NSRegularExpression(pattern: #"<h([1-6])\b[^>]*>(.*?)</h\1>"#, options: [.caseInsensitive, .dotMatchesLineSeparators])
        let matches = pattern.matches(in: source, range: NSRange(source.startIndex..., in: source))
        headings = matches.enumerated().map { index, match in
            var title = (source as NSString).substring(with: match.range(at: 2)).replacingOccurrences(of: "<[^>]+>", with: "", options: .regularExpression)
            for (entity, value) in [("&lt;", "<"), ("&gt;", ">"), ("&quot;", "\""), ("&#39;", "'"), ("&nbsp;", " "), ("&amp;", "&")] { title = title.replacingOccurrences(of: entity, with: value) }
            return Heading(id: "forge-section-\(index)", title: title.trimmingCharacters(in: .whitespacesAndNewlines), level: Int((source as NSString).substring(with: match.range(at: 1))) ?? 1)
        }
        for (index, match) in matches.enumerated().reversed() { source = (source as NSString).replacingCharacters(in: NSRange(location: match.range.location, length: 0), with: "<span id=\"forge-section-\(index)\"></span>") }
        self.html = source
    }

    func imagePath(_ url: URL) -> String? {
        guard url.scheme == "forge-readme", url.host == "image", url.user == nil, url.port == nil, url.query == nil else { return nil }
        let path = String(url.path.dropFirst())
        return RepositoryFile(name: "", path: path, sha: sha, type: "file", size: 0).safePath ? path : nil
    }

    static func escape(_ value: String) -> String { value.replacingOccurrences(of: "&", with: "&amp;").replacingOccurrences(of: "\"", with: "&quot;").replacingOccurrences(of: "<", with: "&lt;").replacingOccurrences(of: ">", with: "&gt;") }
    func page(dark: Bool, outline: Bool = false) -> String {
        let contents = outline && !headings.isEmpty ? "<nav aria-label=\"Contents\"><details open><summary>Contents</summary>" + headings.map { "<p style=\"margin:6px 0 6px \(($0.level - 1) * 12)px\"><a href=\"#\($0.id)\">\(Self.escape($0.title))</a></p>" }.joined() + "</details></nav>" : ""
        return """
        <!doctype html><html><head><meta name="viewport" content="width=device-width,initial-scale=1"><meta http-equiv="Content-Security-Policy" content="default-src 'none'; script-src 'none'; img-src https: forge-readme:; style-src 'unsafe-inline'; base-uri 'none'; form-action 'none'"><style>
        :root{color-scheme:\(dark ? "dark" : "light")}body{margin:0;padding:16px;font:16px -apple-system,BlinkMacSystemFont,Roboto,sans-serif;line-height:1.55;overflow-wrap:anywhere;color:\(dark ? "#e6edf3" : "#1f2328");background:\(dark ? "#0d1117" : "white")}img,video{max-width:100%;height:auto}h1,h2{border-bottom:1px solid \(dark ? "#30363d" : "#d0d7de");padding-bottom:.3em}a{color:\(dark ? "#58a6ff" : "#0969da")}pre{overflow:auto;padding:12px;background:\(dark ? "#161b22" : "#f6f8fa");border-radius:8px}code{font-family:ui-monospace,monospace;font-size:.86em}table{display:block;overflow:auto;border-collapse:collapse}td,th{border:1px solid #8886;padding:6px 12px}blockquote{margin-left:0;border-left:3px solid #8886;padding-left:16px;color:#888}svg{max-width:100%}.anchor{display:none}input{pointer-events:none}
        </style></head><body>\(contents)\(html)</body></html>
        """
    }
}

extension GitHubClient {
    func editProfile(original: GitHubAccount, fields: [String: String], hireable: Bool) async throws {
        let allowed: Set<String> = ["name", "bio", "blog", "company", "location", "twitter_username"]
        guard Set(fields.keys).isSubset(of: allowed), (fields["bio"] ?? "").count <= 160 else { throw GitHubError("Use the supported profile fields and a bio of up to 160 characters.") }
        await clearCache()
        let current = try await profile()
        guard current.login == original.login else { throw GitHubError("The connected account changed. Reopen your profile.") }
        func values(_ user: GitHubAccount) -> [String: String] { ["name": user.name ?? "", "bio": user.bio ?? "", "blog": user.blog ?? "", "company": user.company ?? "", "location": user.location ?? "", "twitter_username": user.twitterUsername ?? ""] }
        let old = values(original), now = values(current)
        var payload: [String: Any] = [:]
        for (key, value) in fields where value != old[key] {
            guard now[key] == old[key] else { throw GitHubError("Your \(key) changed elsewhere. Reopen the editor to avoid overwriting it.") }
            payload[key] = value
        }
        if hireable != (original.hireable ?? false) {
            guard (current.hireable ?? false) == (original.hireable ?? false) else { throw GitHubError("Your availability changed elsewhere. Refresh your profile.") }
            payload["hireable"] = hireable
        }
        if !payload.isEmpty { _ = try await mutationData("/user", method: "PATCH", body: payload) }
    }
    func follows(login: String) async throws -> Bool {
        guard GitHubAccount.validLogin(login), !token.isEmpty else { return false }
        let (_, response) = try await session.data(for: request("/user/following/\(login)"))
        if (response as? HTTPURLResponse)?.statusCode == 404 { return false }; try Self.validate(response); return true
    }
    func follow(login: String, following: Bool) async throws {
        guard GitHubAccount.validLogin(login) else { throw GitHubError("Invalid account.") }
        _ = try await mutationData("/user/following/\(login)", method: following ? "PUT" : "DELETE", body: [:])
    }
    func readme(in repository: Repository, sha: String) async throws -> (RepositoryFile, ReadmeDocument) {
        let query = [URLQueryItem(name: "ref", value: sha)]
        let file: RepositoryFile = try await get("/repos/\(repository.fullName)/readme", query: query)
        guard file.size <= 1_048_576 else { throw GitHubError("This README is over 1 MiB. Open it in Code to download the complete file.") }
        let (data, response) = try await cachedData(for: request("/repos/\(repository.fullName)/readme", query: query, accept: "application/vnd.github.html+json"))
        try Self.validate(response)
        guard data.count <= 4_194_304 else { throw GitHubError("This rendered README is too large to preview.") }
        return (file, try ReadmeDocument(html: String(decoding: data, as: UTF8.self), repository: repository, sha: sha, path: file.path))
    }

    func readmeImage(_ url: URL, document: ReadmeDocument) async throws -> Data {
        guard let path = document.imagePath(url) else { throw GitHubError("Unsupported README image.") }
        let request = try request("/repos/\(document.repository.fullName)/contents/\(path)", query: [URLQueryItem(name: "ref", value: document.sha)], accept: "application/vnd.github.raw+json")
        let epoch = await cache?.epoch ?? 0
        if let entry = await cache?.value(for: request), entry.fresh { return entry.data }
        let (file, response) = try await session.download(for: request)
        defer { try? FileManager.default.removeItem(at: file) }; try Self.validate(response)
        let handle = try FileHandle(forReadingFrom: file); defer { try? handle.close() }
        let data = try handle.read(upToCount: 8_388_609) ?? Data()
        guard data.count <= 8_388_608 else { throw GitHubError("README images must be under 8 MiB.") }
        if let response = response as? HTTPURLResponse { await cache?.store(data, response: response, for: request, epoch: epoch) }; return data
    }

    func profileHighlights(login: String) async throws -> ProfileHighlights {
        guard GitHubAccount.validLogin(login) else { throw GitHubError("Invalid profile.") }
        struct Result: Decodable { let user: ProfileHighlights? }
        let result: Result = try await graphQL("query($login:String!){user(login:$login){isEmployee isDeveloperProgramMember isGitHubStar isCampusExpert pinnedItems(first:6,types:[REPOSITORY]){nodes{... on Repository{nameWithOwner description stargazerCount}}}}}", variables: ["login": login])
        guard let user = result.user else { throw GitHubError("Pinned repositories are unavailable for this account.") }; return user
    }

    func setStar(in repository: Repository, starred: Bool) async throws {
        _ = try await mutationData("/user/starred/\(repository.fullName)", method: starred ? "PUT" : "DELETE", body: [:])
    }
    func isStarred(_ repository: Repository) async throws -> Bool {
        guard !token.isEmpty else { return false }
        let (_, response) = try await session.data(for: request("/user/starred/\(repository.fullName)"))
        if (response as? HTTPURLResponse)?.statusCode == 404 { return false }; try Self.validate(response); return true
    }
    func editDescription(in repository: Repository, original: String, description: String) async throws {
        await clearCache()
        let current: RepositoryOverview = try await get("/repos/\(repository.fullName)")
        guard current.permissions?.admin == true, (current.description ?? "") == original else { throw GitHubError("Only an administrator can edit the description. Refresh if it changed elsewhere.") }
        _ = try await mutationData("/repos/\(repository.fullName)", method: "PATCH", body: ["description": description])
    }

    func issueOptions(in repository: Repository, kind: String, page: Int, cursor: String? = nil) async throws -> IssueOptionPage {
        let prefix = "/repos/\(repository.fullName)"
        switch kind {
        case "Assignees":
            let rows: [GitHubAccount] = try await get(prefix + "/assignees", page: page, count: 30)
            return IssueOptionPage(items: rows.map { IssueOption(id: $0.login, title: $0.login) }, more: rows.count == 30)
        case "Labels":
            struct Label: Decodable { let name: String; let description: String? }
            let rows: [Label] = try await get(prefix + "/labels", page: page, count: 30)
            return IssueOptionPage(items: rows.map { IssueOption(id: $0.name, title: $0.name, detail: $0.description ?? "") }, more: rows.count == 30)
        case "Milestone":
            struct Milestone: Decodable { let number: Int; let title: String; let description: String? }
            let rows: [Milestone] = try await get(prefix + "/milestones", page: page, count: 30, query: [URLQueryItem(name: "state", value: "open")])
            return IssueOptionPage(items: rows.map { IssueOption(id: String($0.number), title: $0.title, detail: $0.description ?? "") }, more: rows.count == 30)
        case "Project":
            struct Project: Decodable { let id: String; let title: String; let closed: Bool }
            struct Result: Decodable { struct Repo: Decodable { let projectsV2: GraphQLConnection<Project> }; let repository: Repo? }
            let parts = repository.fullName.split(separator: "/").map(String.init)
            let result: Result = try await graphQL("query($owner:String!,$name:String!,$cursor:String){repository(owner:$owner,name:$name){projectsV2(first:30,after:$cursor){nodes{id title closed}pageInfo{hasNextPage endCursor}}}}", variables: ["owner": parts[0], "name": parts[1], "cursor": cursor as Any? ?? NSNull()])
            guard let connection = result.repository?.projectsV2 else { throw GitHubError("No linked projects are accessible. Check Projects permission.") }
            return IssueOptionPage(items: connection.nodes.compactMap { $0 }.filter { !$0.closed }.map { IssueOption(id: $0.id, title: $0.title) }, more: connection.pageInfo.hasNextPage, cursor: connection.pageInfo.endCursor)
        default: throw GitHubError("Unknown issue field.")
        }
    }
    func createIssue(in repository: Repository, fields: IssueFields) async throws -> (Conversation, String?) {
        let data = try await mutationData("/repos/\(repository.fullName)/issues", body: fields.payload())
        let issue = try Self.decoder().decode(Conversation.self, from: data)
        let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
        let labels = (object["labels"] as? [[String: Any]] ?? []).compactMap { $0["name"] as? String }
        let assignees = (object["assignees"] as? [[String: Any]] ?? []).compactMap { $0["login"] as? String }
        let milestone = (object["milestone"] as? [String: Any])?["number"] as? Int
        let missing = !Set(fields.labels).isSubset(of: Set(labels)) || !Set(fields.assignees).isSubset(of: Set(assignees)) || (fields.milestone != nil && fields.milestone != milestone)
        return (issue, missing ? "The issue was created, but GitHub did not apply all selected metadata. Check your repository and token permissions." : nil)
    }
    func addIssueToProject(issueID: String, projectID: String) async throws {
        struct Result: Decodable { struct Added: Decodable { struct Item: Decodable { let id: String }; let item: Item? }; let addProjectV2ItemById: Added? }
        let result: Result = try await graphQL("mutation($project:ID!,$issue:ID!){addProjectV2ItemById(input:{projectId:$project,contentId:$issue}){item{id}}}", variables: ["project": projectID, "issue": issueID])
        guard result.addProjectV2ItemById?.item != nil else { throw GitHubError("GitHub did not confirm the project assignment.") }
    }
}

File: Android/app/src/main/java/app/forge/github/RepositoryExperience.kt
package app.forge.github

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.URI

fun issueFields(title: String, body: String, assignees: List<String> = emptyList(), labels: List<String> = emptyList(), milestone: Int? = null): JSONObject {
    require(title.trim().isNotEmpty() && title.trim().length <= 256 && assignees.all(::validLogin) && assignees.size <= 10 && (milestone == null || milestone > 0)) { "Enter a title and valid issue metadata." }
    return json("title" to title.trim(), "body" to body).apply {
        if (assignees.isNotEmpty()) put("assignees", JSONArray(assignees))
        if (labels.isNotEmpty()) put("labels", JSONArray(labels))
        milestone?.let { put("milestone", it) }
    }
}

fun htmlEscape(value: String) = value.replace("&", "&amp;").replace("\"", "&quot;").replace("<", "&lt;").replace(">", "&gt;")

class ReadmeDocument(source: String, val repo: String, val sha: String, val path: String) {
    val html: String
    val base: String get() = URI("https", "github.com", "/$repo/blob/$sha/$path", null).toASCIIString()
    init {
        repository(repo); require(sha.matches(Regex("[a-fA-F0-9]{40}"))); require(safePath(path))
        val base = URI("https", "raw.githubusercontent.com", "/$repo/$sha/$path", null)
        val clean = source.replace(Regex("""\s+srcset\s*=\s*(?:"[^"]*"|'[^']*')""", RegexOption.IGNORE_CASE), "")
        html = Regex("""\bsrc\s*=\s*(["'])(.*?)\1""", setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL)).replace(clean) { match ->
            val location = runCatching {
                val uri = base.resolve(match.groupValues[2].replace("&amp;", "&")).normalize()
                require(uri.scheme == "https" && uri.userInfo == null && uri.port in listOf(-1, 443))
                val rawPrefix = "/$repo/$sha/"; val webPrefix = "/$repo/raw/$sha/"
                val image = when {
                    uri.host == "raw.githubusercontent.com" && uri.path.startsWith(rawPrefix) -> uri.path.removePrefix(rawPrefix)
                    uri.host == "github.com" && uri.path.startsWith(webPrefix) -> uri.path.removePrefix(webPrefix)
                    else -> null
                }
                if (image != null) URI("forge-readme", "image", "/$image", null).toASCIIString() else uri.toASCIIString()
            }.getOrDefault("")
            "src=\"${htmlEscape(location)}\""
        }
    }
    fun imagePath(uri: URI): String? = runCatching {
        require(uri.scheme == "forge-readme" && uri.host == "image" && uri.userInfo == null && uri.port == -1 && uri.query == null)
        uri.path.removePrefix("/").also { require(safePath(it)) }
    }.getOrNull()
    fun page(dark: Boolean): String = """<!doctype html><html><head><meta name="viewport" content="width=device-width,initial-scale=1"><meta http-equiv="Content-Security-Policy" content="default-src 'none'; script-src 'none'; img-src https: forge-readme:; style-src 'unsafe-inline'; base-uri 'none'; form-action 'none'"><style>
        :root{color-scheme:${if(dark) "dark" else "light"}}body{margin:0;padding:16px;font:16px -apple-system,BlinkMacSystemFont,Roboto,sans-serif;line-height:1.55;overflow-wrap:anywhere;color:${if(dark) "#e6edf3" else "#1f2328"};background:${if(dark) "#0d1117" else "white"}}img,video{max-width:100%;height:auto}h1,h2{border-bottom:1px solid #8886;padding-bottom:.3em}a{color:${if(dark) "#58a6ff" else "#0969da"}}pre{overflow:auto;padding:12px;background:${if(dark) "#161b22" else "#f6f8fa"};border-radius:8px}code{font-family:ui-monospace,monospace;font-size:.86em}table{display:block;overflow:auto;border-collapse:collapse}td,th{border:1px solid #8886;padding:6px 12px}blockquote{margin-left:0;border-left:3px solid #8886;padding-left:16px;color:#888}svg{max-width:100%}.anchor{display:none}input{pointer-events:none}
        </style></head><body>$html</body></html>"""
}

suspend fun GitHub.data(path: String, query: Map<String, String> = emptyMap(), accept: String = "application/vnd.github.raw+json", limit: Int = 8_388_608): ByteArray = withContext(Dispatchers.IO) {
    var uri = apiUrl(path, query)
    val key = "$token\n$uri\n$accept"; val epoch = cache?.epoch ?: 0; val old = cache?.value(key)
    if (old?.fresh == true) return@withContext old.bytes
    repeat(6) {
        val c = connection(uri, accept)
        old?.etag?.let { c.setRequestProperty("If-None-Match", it) }
        try {
            if (c.responseCode == 304 && old != null) { cache?.store(key, old.bytes, c.getHeaderField("ETag") ?: old.etag, epoch, c.getHeaderField("Cache-Control") ?: old.control); return@withContext old.bytes }
            if (c.responseCode in listOf(301, 302, 307, 308)) {
                val next = uri.resolve(c.getHeaderField("Location") ?: error("Missing redirect."))
                require(next.host == "api.github.com" && trustedDownload(next)) { "Unexpected API redirect." }; uri = next
            } else {
                require(c.responseCode in 200..299) { "GitHub returned HTTP ${c.responseCode}. Check access or retry." }
                return@withContext c.inputStream.use { it.readLimited(limit + 1) }.also {
                    require(it.size <= limit) { "This preview is too large. Download the file instead." }
                    cache?.store(key, it, c.getHeaderField("ETag"), epoch, c.getHeaderField("Cache-Control") ?: "")
                }
            }
        } finally { c.disconnect() }
    }
    error("Too many redirects.")
}

suspend fun GitHub.readme(repo: String, sha: String): Pair<JSONObject, ReadmeDocument> {
    repository(repo); require(validSha(sha))
    val query = mapOf("ref" to sha); val file = obj("/repos/$repo/readme", query)
    require(file.optLong("size") <= 1_048_576) { "This README is over 1 MiB. Open Code to download the full file." }
    val html = data("/repos/$repo/readme", query, "application/vnd.github.html+json", 4_194_304).toString(Charsets.UTF_8)
    return file to ReadmeDocument(html, repo, sha, file.s("path"))
}

suspend fun GitHub.profileHighlights(login: String) = gql("""query(${'$'}login:String!){user(login:${'$'}login){isEmployee isDeveloperProgramMember isGitHubStar isCampusExpert pinnedItems(first:6,types:[REPOSITORY]){nodes{... on Repository{nameWithOwner description stargazerCount}}}}}""", json("login" to accountPath(login))).o("user")

suspend fun GitHub.isStarred(repo: String): Boolean = withContext(Dispatchers.IO) {
    if (token.isBlank()) return@withContext false
    val c = connection(apiUrl("/user/starred/${repository(repo)}"))
    try { if (c.responseCode == 404) false else { require(c.responseCode == 204) { "Couldn't load your star status." }; true } } finally { c.disconnect() }
}

suspend fun GitHub.editDescription(repo: String, original: String, description: String) {
    cache?.clear()
    val current = obj("/repos/${repository(repo)}")
    require(current.o("permissions").optBoolean("admin") && current.s("description") == original) { "Only administrators can edit descriptions. Refresh if it changed elsewhere." }
    change("/repos/$repo", "PATCH", json("description" to description))
}

data class IssueOption(val id: String, val title: String, val detail: String = "")
data class IssueOptionPage(val items: List<IssueOption>, val more: Boolean, val cursor: String = "")

suspend fun GitHub.issueOptions(repo: String, kind: String, page: Int, cursor: String): IssueOptionPage {
    val prefix = "/repos/${repository(repo)}"
    if (kind == "Project") {
        val parts = repo.split('/')
        val connection = gql("""query(${'$'}owner:String!,${'$'}name:String!,${'$'}cursor:String){repository(owner:${'$'}owner,name:${'$'}name){projectsV2(first:30,after:${'$'}cursor){nodes{id title closed}pageInfo{hasNextPage endCursor}}}}""", json("owner" to parts[0], "name" to parts[1], "cursor" to cursor.ifBlank { null })).o("repository").getJSONObject("projectsV2")
        return IssueOptionPage(connection.rows("nodes").filter { !it.optBoolean("closed") }.map { IssueOption(it.s("id"), it.s("title")) }, connection.o("pageInfo").optBoolean("hasNextPage"), connection.o("pageInfo").s("endCursor"))
    }
    val route = when(kind) { "Assignees" -> "assignees"; "Labels" -> "labels"; "Milestone" -> "milestones"; else -> error("Unknown issue field.") }
    val rows = list("$prefix/$route", page, if (kind == "Milestone") mapOf("state" to "open") else emptyMap())
    return IssueOptionPage(rows.map { row -> when(kind) {
        "Assignees" -> IssueOption(row.s("login"), row.s("login"))
        "Labels" -> IssueOption(row.s("name"), row.s("name"), row.s("description"))
        else -> IssueOption(row.s("number"), row.s("title"), row.s("description"))
    } }, rows.size == 30)
}

suspend fun GitHub.addIssueToProject(issueID: String, projectID: String) {
    val result = gql("""mutation(${'$'}project:ID!,${'$'}issue:ID!){addProjectV2ItemById(input:{projectId:${'$'}project,contentId:${'$'}issue}){item{id}}}""", json("project" to projectID, "issue" to issueID))
    require(result.o("addProjectV2ItemById").o("item").s("id").isNotBlank()) { "GitHub did not confirm project assignment." }
}

suspend fun GitHub.editProfile(original: JSONObject, fields: Map<String, String>, hireable: Boolean) {
    val allowed = setOf("name", "bio", "blog", "company", "location", "twitter_username")
    require(fields.keys.all { it in allowed } && (fields["bio"]?.length ?: 0) <= 160) { "Keep the bio within 160 characters." }
    cache?.clear()
    val current = obj("/user"); require(current.s("login") == original.s("login")) { "Your account changed. Reopen the editor." }
    val changes = JSONObject()
    fields.forEach { (key, value) -> if (value != original.s(key)) {
        require(current.s(key) == original.s(key)) { "Your profile changed elsewhere. Refresh before saving." }; changes.put(key, value)
    } }
    if (hireable != original.optBoolean("hireable")) {
        require(current.optBoolean("hireable") == original.optBoolean("hireable")) { "Your profile changed elsewhere. Refresh before saving." }
        changes.put("hireable", hireable)
    }
    if (changes.length() > 0) change("/user", "PATCH", changes)
}

suspend fun GitHub.isFollowing(login: String): Boolean = withContext(Dispatchers.IO) {
    val c = connection(apiUrl("/user/following/${accountPath(login)}"))
    try { if (c.responseCode == 404) false else { require(c.responseCode == 204) { "Couldn't load your follow status." }; true } } finally { c.disconnect() }
}
